scylla-cql/src/types/deserialize/result.rs

```diff
@@ -132,46 +136,212 @@ where
+// and the std Iterator interface does not allow for that.
+/// A _lending_ iterator over serialized rows.
+///
+/// This type is similar to `RowIterator`, but keeps ownership of the serialized
+/// result. Because it returns `ColumnIterator`s that need to borrow from it,
+/// it does not implement the `Iterator` trait (there is no type in the standard
+/// library to represent this concept yet).
+#[derive(Debug)]
+pub struct RawRowsLendingIterator {
+    metadata: ResultMetadataHolder<'static>,
+    remaining: usize,
+    at: usize,
+    raw_rows: Bytes,
+}
+
+impl RawRowsLendingIterator {
+    /// Creates a new `RawRowsLendingIterator`, consuming given `RawRows`.
+    #[inline]
+    pub fn new(raw_rows: RawRowsWithDeserializedMetadata<'static>) -> Self {
+        let (metadata, rows_count, raw_rows) = raw_rows.into_inner();
+        Self {
+            metadata,
+            remaining: rows_count,
+            at: 0,
+            raw_rows,
+        }
+    }
+
+    /// Returns a `ColumnIterator` that represents the next row.
+    ///
+    /// Note: the `ColumnIterator` borrows from the `RawRowsLendingIterator`.
+    /// The column iterator must be consumed before the rows iterator can
+    /// continue.
+    #[inline]
+    #[allow(clippy::should_implement_trait)] // https://github.com/rust-lang/rust-clippy/issues/5004
+    pub fn next(&mut self) -> Option<Result<ColumnIterator, DeserializationError>> {
+        self.remaining = self.remaining.checked_sub(1)?;
+
+        // First create the slice encompassing the whole frame.
+        let mut remaining_frame = FrameSlice::new(&self.raw_rows);
+        // Then slice it to encompass the remaining suffix of the frame.
+        *remaining_frame.as_slice_mut() = &remaining_frame.as_slice()[self.at..];
+        // Ideally, we would prefer to preserve the FrameSlice between calls to `next()`,
+        // but borrowing from oneself is impossible, so we have to recreate it this way.
+
+        let iter = ColumnIterator::new(self.metadata.col_specs(), remaining_frame);
+
+        // Skip the row here, manually
+        for (column_index, spec) in self.metadata.col_specs().iter().enumerate() {
+            let remaining_frame_len_before_column_read = remaining_frame.as_slice().len();
+            if let Err(err) = remaining_frame.read_cql_bytes() {
+                return Some(Err(mk_deser_err::<Self>(
+                    BuiltinDeserializationErrorKind::RawColumnDeserializationFailed {
+                        column_index,
+                        column_name: spec.name().to_owned(),
+                        err: DeserializationError::new(err),
+                    },
+                )));
+            } else {
+                let remaining_frame_len_after_column_read = remaining_frame.as_slice().len();
+                self.at +=
+                    remaining_frame_len_before_column_read - remaining_frame_len_after_column_read;
+            }
+        }
```

Comment:
Correct. But *reading* here means just skipping over a CQL bytes object.

Reply:
Which could still be a problem if there is a lot of columns, right?

Reply:
This indeed doubles the overhead of going through the structure of serialized data. Is it a problem? Perhaps.

The problem arises from the CQL protocol: it does not specify the length of the row. In order to pass over a row, one has to go through all the columns (which, fortunately, have their length provided).

I thought about some solution based on `ColumnIterator` mutably borrowing `RowIterator`'s FrameSlice`; then all changes to it done by `ColumnIterator` would be already present on `RowIterator`. This could be implemented by parametrising `ColumnIterator` with a third lifetime (`'row_iter`).
It does not always work, however. `ColumnIterator` could be dropped without completing iteration, and this way `RowIterator` would have its `FrameSlice` pointing at invalid place in memory (somewhere inside a row, not at the beginning of a row). Just in case you thought now about finishing `ColumnIterator`'s iteration in its custom `Drop` impl, then forget about this. _Safe Rust_ allows `mem::forget()` and thus we cannot rely on destructors being run.

Reply:
In an offline discussion with @Lorak-mmk we decided that this needs to be left as-is.

Reply:
The described solution would require `RowIterator` to be a lending iterator, which is a major drawback for the user-facing API (`Iterator` trait would not be viable).